src/mds/Server.cc

```diff
@@ -4377,12 +4377,12 @@ void Server::handle_client_open(MDRequestRef& mdr)
     respond_to_request(mdr, -CEPHFS_EROFS);
     return;
   }
-  
+
   CInode *cur = rdlock_path_pin_ref(mdr, need_auth);
   if (!cur)
     return;
 
-  if (cur->is_frozen() || cur->state_test(CInode::STATE_EXPORTINGCAPS)) {
+  if ((cur->is_frozen() || cur->state_test(CInode::STATE_EXPORTINGCAPS)) && !need_auth) {
```

Comment:
So, the request should have been forwarded if it was not the auth mds and then we won't reach here. This looks fine.

Reply:
Yeah, correct.

Reply:
Ok, but thinking more on this - it requires multiple MDSs to run into the deadlock? If yes, I'm not sure how this was hit in the setup where the deadlock was seen.

Reply:
I don't think this can be it because the inode dump does not show it as frozen or exporting caps.

Reply:
> Ok, but thinking more on this - it requires multiple MDSs to run into the deadlock? If yes, I'm not sure how this was hit in the setup where the deadlock was seen.

There were 2 active MDSs.

Reply:
> I don't think this can be it because the inode dump does not show it as frozen or exporting caps.

I didn't notice this caps. This is the only case I can figure out in the create case could cause the deadlock till now.

Reply:
I think @batrick is correct, it should be the `filelock` instead of dentry's lock:

```
    "filelock": {
        "gather_set": [],
        "state": "xlock",
        "is_leased": false,
        "num_rdlocks": 0,
        "num_wrlocks": 0,
        "num_xlocks": 1,
        "xlock_by": {
            "flag_point": "failed to rdlock, waiting",
            "reqid": "client.161452256:598076",
            "op_type": "client_request",
            "client_info": {
                "client": "client.161452256",
                "tid": 598076
            },
            "events": [
                {
                    "time": "2023-07-13T09:56:49.740519+0000",
                    "event": "initiated"
                },
...
                {
                    "time": "2023-07-13T09:56:49.835378+0000",
                    "event": "failed to xlock, waiting"
                },
                {
                    "time": "2023-07-13T09:56:50.150399+0000",
                    "event": "acquired locks"
                },
                {
                    "time": "2023-07-13T09:56:50.150438+0000",
                    "event": "failed to rdlock, waiting"
                }
            ]
        }
```

Currently for the `create` request there is only one case will `xlock` the `filelock` in `handle_client_open()` in truncate case. But after that I haven't find any other place is trying to acquire the `rdlock` for `filelock` yet.

Reply:
> > Ok, but thinking more on this - it requires multiple MDSs to run into the deadlock? If yes, I'm not sure how this was hit in the setup where the deadlock was seen.
> 
> There were 2 active MDSs.

Yeh, but the inode wasn't frozen, so it doesn't look to be this call path for the deadlock.

Reply:
Sure.

There have two case:

1, `_openc()` -->` _open()`: the first call of `rdlock_path_pin_ref()` will do nothing even the `need_auth == true`, because in `_openc()` the `rdlock_path_xlock_dentry()` will set the `MutationImpl::PATH_LOCKED` flag.

So we should retry the `rdlock_path_pin_ref()` anyway if `need_auth == true` with:

```
-  if (cur->is_frozen() || cur->state_test(CInode::STATE_EXPORTINGCAPS)) {
-    ceph_assert(!need_auth);
+  if (need_auth || cur->is_frozen() || cur->state_test(CInode::STATE_EXPORTINGCAPS)) {
     mdr->locking_state &= ~(MutationImpl::PATH_LOCKED | MutationImpl::ALL_LOCKED);
     CInode *cur = rdlock_path_pin_ref(mdr, true);
     if (!cur)
```

2, `_open()` directly: the first call of `rdlock_path_pin_ref()` will forward to auth directly if `need_auth == true`, so it shouldn't be here.

Reply:
> So we should retry the rdlock_path_pin_ref() anyway if need_auth == true with:

That will always rerun `rdlock_path_pin_ref` twice if `need_auth == true` though, yes? I think we should try to avoid that except in the case of `openc -> open`. I wonder why not `mdr->locking_state &= ~(MutationImpl::PATH_LOCKED | MutationImpl::ALL_LOCKED);` before having `openc` call `open`.

Reply:
> > So we should retry the rdlock_path_pin_ref() anyway if need_auth == true with:
> 
> That will always rerun `rdlock_path_pin_ref` twice if `need_auth == true` though, yes? I think we should try to avoid that except in the case of `openc -> open`. I wonder why not `mdr->locking_state &= ~(MutationImpl::PATH_LOCKED | MutationImpl::ALL_LOCKED);` before having `openc` call `open`.

No, not always. Mostly the first one will do nothing in case of `_openc()` --> `_open()` and no need to do anything. Because the clients will always try to send the `WRITE` request to auth MDS and the `_openc()` has already done what we need here.

Just in case the open requests need be handled in auth `MDS` but the current `MDS` is not auth or the inode is under exporting will we need to retry it.